streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,229 @@
+        return false;
+    }
+
+    @Override
+    public boolean isAllowedTaskMovement(final ClientState source, final ClientState destination) {
+        final Map<String, String> sourceClientTags = source.clientTags();
+        final Map<String, String> destinationClientTags = destination.clientTags();
+
+        for (final Entry<String, String> sourceClientTagEntry : sourceClientTags.entrySet()) {
+            if (!sourceClientTagEntry.getValue().equals(destinationClientTags.get(sourceClientTagEntry.getKey()))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static void fillClientsTagStatistics(final Map<UUID, ClientState> clientStates,
+                                                 final Map<TagEntry, Set<UUID>> tagEntryToClients,
+                                                 final Map<String, Set<String>> tagKeyToValues) {
+        for (final Entry<UUID, ClientState> clientStateEntry : clientStates.entrySet()) {
+            final UUID clientId = clientStateEntry.getKey();
+            final ClientState clientState = clientStateEntry.getValue();
+
+            clientState.clientTags().forEach((tagKey, tagValue) -> {
+                tagKeyToValues.computeIfAbsent(tagKey, ignored -> new HashSet<>()).add(tagValue);
+                tagEntryToClients.computeIfAbsent(new TagEntry(tagKey, tagValue), ignored -> new HashSet<>()).add(clientId);
+            });
+        }
+    }
+
+    private static void assignStandbyTasksForActiveTask(final int numStandbyReplicas,
+                                                        final TaskId activeTaskId,
+                                                        final UUID activeTaskClient,
+                                                        final Set<String> rackAwareAssignmentTags,
+                                                        final Map<UUID, ClientState> clientStates,
+                                                        final Map<TaskId, Integer> tasksToRemainingStandbys,
+                                                        final Map<String, Set<String>> tagKeyToValues,
+                                                        final Map<TagEntry, Set<UUID>> tagEntryToClients) {
+        final ConstrainedPrioritySet standbyTaskClientsByTaskLoad = new ConstrainedPrioritySet(
+            (client, t) -> !clientStates.get(client).hasAssignedTask(t),
+            client -> clientStates.get(client).assignedTaskLoad()
+        );
+
+        final Set<UUID> usedClients = new HashSet<>();
+
+        standbyTaskClientsByTaskLoad.offerAll(clientStates.keySet());
+
+        int numRemainingStandbys = tasksToRemainingStandbys.get(activeTaskId);
+
+        usedClients.add(activeTaskClient);
+
+        while (numRemainingStandbys > 0) {
+            final Set<UUID> clientsOnAlreadyUsedTagDimensions = findClientsOnUsedTagDimensions(
```

Comment:
Fixed with e3aff39.
Now we only create `clientsOnAlreadyUsedTagDimensions` once and populate it for the each standby task assignment instead of re-creating it.